perun-core/src/main/java/cz/metacentrum/perun/core/impl/Utils.java

```diff
@@ -1009,84 +1007,106 @@ public static String getMessageAuthenticationCode(String input) {
 	 * @throws InternalErrorException
 	 */
 	public static void sendValidationEmail(User user, String url, String email, int changeId, String subject, String content, String customUrlPath) {
-
-		JavaMailSender mailSender = BeansUtils.getDefaultMailSender();
-
-		// create message
-		SimpleMailMessage message = new SimpleMailMessage();
-		message.setTo(email);
-		message.setFrom(BeansUtils.getCoreConfig().getMailchangeBackupFrom());
-
 		String instanceName = BeansUtils.getCoreConfig().getInstanceName();
 
-		if (subject == null ||subject.isEmpty()) {
-			message.setSubject("["+instanceName+"] New email address verification");
-		} else {
-			subject = subject.replace("{instanceName}", instanceName);
-			message.setSubject(subject);
-		}
-
-		// get validation link params
-		String i = Integer.toString(changeId, Character.MAX_RADIX);
-		String m = Utils.getMessageAuthenticationCode(i);
-
+		// use default if unknown rpc path
+		String linkLocation = "/gui/";
 		try {
-
-			// !! There is a hard-requirement for Perun instance
-			// to host GUI on same server as RPC like: "serverUrl/gui/"
-
 			URL urlObject = new URL(url);
-
-			// use default if unknown rpc path
-			String path = "/gui/";
-
-			if (customUrlPath != null) {
-				path = customUrlPath;
+			//if there is custom path, use it, if not, try to use specific path
+			if (customUrlPath != null && !customUrlPath.isEmpty()) {
+				linkLocation = customUrlPath;
 			} else if (urlObject.getPath().contains("/krb/")) {
-				path = "/krb/gui/";
+				linkLocation = "/krb/gui/";
 			} else if (urlObject.getPath().contains("/fed/")) {
-				path = "/fed/gui/";
+				linkLocation = "/fed/gui/";
 			} else if (urlObject.getPath().contains("/ceitec/")) {
 				// to support ceitec proxy, since it gets more fed attributes than IDM SP.
-				path = "/ceitec/gui/";
+				linkLocation = "/ceitec/gui/";
 			} else if (urlObject.getPath().contains("/cert/")) {
-				path = "/cert/gui/";
+				linkLocation = "/cert/gui/";
 			}
+		} catch (MalformedURLException ex) {
+			throw new InternalErrorException("Not valid URL of running Perun instance.", ex);
+		}
+		String validationLink = prepareValidationLinkForEmailChange(url, linkLocation, changeId, user);
```

Comment:
``validationLink`` is unused in the later code. Maybe it get mixed with ``linkLocation`` in the template?

Reply:
Yes, I used linkLocation instead of validation link. It should be fixed now.

Reply:
It remained in ``defaultBody`` as directly joined strings. Maybe you should just prepare all replacable strings first and then use placeholders (eg. ``{link}``) also in default template body. This might apply to other mail types too.

Reply:
You are right, it is still in the body. I have replaced that.
If you are using defaultBody or defaultSubject I would prefer to not doing replacements. It is better to have it ready before using.